contracts/cw-pre-propose-base-proposal-single/src/contract.rs

```diff
@@ -0,0 +1,295 @@
+#[cfg(not(feature = "library"))]
+use cosmwasm_std::entry_point;
+use cosmwasm_std::{
+    Binary, CosmosMsg, Deps, DepsMut, Empty, Env, MessageInfo, Response, StdResult,
+};
+use cw2::set_contract_version;
+
+use cw_pre_propose_base::{
+    error::PreProposeError,
+    msg::{ExecuteMsg as ExecuteBase, InstantiateMsg as InstantiateBase, QueryMsg as QueryBase},
+    state::PreProposeContract,
+};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+const CONTRACT_NAME: &str = "crates.io:cw-pre-propose-base-proposal-single";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[derive(Serialize, JsonSchema, Deserialize, Debug, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum ProposeMessage {
+    Propose {
+        title: String,
+        description: String,
+        msgs: Vec<CosmosMsg<Empty>>,
+    },
+}
+
+pub type InstantiateMsg = InstantiateBase<Empty>;
+pub type ExecuteMsg = ExecuteBase<ProposeMessage, Empty>;
+pub type QueryMsg = QueryBase<Empty>;
+
+/// Internal version of the propose message that includes the
+/// `proposer` field. The module will fill this in based on the sender
+/// of the external message.
+#[derive(Serialize, JsonSchema, Deserialize, Debug, Clone)]
+#[serde(rename_all = "snake_case")]
+enum ProposeMessageInternal {
+    Propose {
+        title: String,
+        description: String,
+        msgs: Vec<CosmosMsg<Empty>>,
+        proposer: Option<String>,
+    },
+}
+
+type PrePropose = PreProposeContract<Empty, Empty, Empty, ProposeMessageInternal>;
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, PreProposeError> {
+    let resp = PrePropose::default().instantiate(deps.branch(), env, info, msg)?;
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    Ok(resp)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, PreProposeError> {
+    type ExecuteInternal = ExecuteBase<ProposeMessageInternal, Empty>;
+    let internalized = match msg {
```

Comment:
Can we add a comment here? This pattern seems to deviate a bit from what CW devs are use to.

Reply:
yes :) this is a little weird. spent a while coming up with this.